### ISSUE TITLE
False positive readability-redundant-declaration warning on template friend of a template class

### ISSUE THREAD

**Issue by yeputons** [llvm/llvm-project#47430]
|  |  |
| --- | --- |
| Bugzilla Link | [48086](https://llvm.org/bz48086) |
| Version | unspecified |
| OS | Windows NT |
| CC | @EugeneZelenko,@fabianwolff |

## Extended Description 
Almost minimal example:

```
template<typename T> class Foo {
    int x{};
    template <typename U> friend int bar(Foo<U> const &); // (1)
};
template<typename U> int bar(Foo<U> const &f) {
    return f.x; // (2)
}
int main() {
    [[maybe_unused]] Foo<int> f;
    bar(f);
}
```

The following command
```
clang-tidy --checks=readability-redundant-declaration a.cpp
```
produces the following result
```
a.cpp:3:38: warning: redundant 'bar' declaration [readability-redundant-declaration]
    template <typename U> friend int bar(Foo<U> const &); // (1)
                          ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
a.cpp:5:26: note: previously declared here
template<typename U> int bar(Foo<U> const &f) {
                         ^
```

which is wrong. If you comment out (1), you get a compilation error because (2) can no longer access the private field because it's not a friend.

My clang-tidy version:

LLVM (http://llvm.org/):
  LLVM version 10.0.0
  Optimized build.
  Default target: x86_64-pc-windows-msvc
  Host CPU: haswell

**Comment by barometz** [comment]
Still present in 13.0.0. Wasn't able to test with 14, unfortunately.

**Comment by PiotrZSL** [comment]
This were fixed by 42bc3275d3687524ddc0d20c72722b9324f87be4.
